Mathlib/Data/Nat/PartENat.lean

```diff
@@ -0,0 +1,850 @@
+    toWithTop 0 = 0 :=
+  rfl
+#align part_enat.to_with_top_zero PartENat.toWithTop_zero
+
+@[simp]
+theorem toWithTop_zero' {h : Decidable (0 : PartENat).Dom} : toWithTop 0 = 0 := by
+  convert toWithTop_zero
+#align part_enat.to_with_top_zero' PartENat.toWithTop_zero'
+
+theorem toWithTop_some (n : ℕ) : toWithTop (some n) = n :=
+  rfl
+#align part_enat.to_with_top_some PartENat.toWithTop_some
+
+theorem toWithTop_natCast (n : ℕ) {_ : Decidable (n : PartENat).Dom} : toWithTop n = n := by
+  simp only [← some_eq_natCast, ← toWithTop_some]
+  congr
+#align part_enat.to_with_top_coe PartENat.toWithTop_natCast
+
+@[simp]
+theorem toWithTop_natCast' (n : ℕ) {h : Decidable (n : PartENat).Dom} :
+    toWithTop (n : PartENat) = n := by
+  rw [toWithTop_natCast n]
+#align part_enat.to_with_top_coe' PartENat.toWithTop_natCast'
+
+-- Porting note: statement changed. Mathlib 3 statement was
+-- ```
+-- @[simp] lemma to_with_top_le {x y : part_enat} :
+--   Π [decidable x.dom] [decidable y.dom], by exactI to_with_top x ≤ to_with_top y ↔ x ≤ y :=
+-- ```
+--
+-- In addition, this proof takes really long to type-check.
+-- This boils down to the manually defined instances in `Data.ENat.Basic`
+-- (see porting note l.46 'Porting note: instances that derive failed to find')
+-- it seems that the last `simp` in the following proof needs to check
+-- `CharZero ℕ∞ ≟ CharZero ℕ∞` where one side uses
+-- `#check instENatAddCommMonoidWithOne`
+-- and the other `NonAssocSemiring.toAddCommMonoidWithOne` and then eventually
+-- `#check instENatCanonicallyOrderedCommSemiring`
+--
+-- See
+-- ```
+-- set_option trace.Meta.synthInstance true
+-- set_option trace.Meta.isDefEq true
+-- set_option profiler true
+-- ```
+@[simp]
+theorem toWithTop_le {x y : PartENat} [hx : Decidable x.Dom] [hy : Decidable y.Dom] :
+    toWithTop x ≤ toWithTop y ↔ x ≤ y := by
+  induction y using PartENat.casesOn generalizing hy
+  · simp
+  induction x using PartENat.casesOn generalizing hx
+  · simp
+  · simp -- Porting note: this takes too long.
+#align part_enat.to_with_top_le PartENat.toWithTop_le
+
+-- Porting note: As part of the troubles above, I noticed that Lean4 does not
+-- find the following two instances which it could find in Lean3 automatically:
+-- ```
+-- section Test
+-- #synth Decidable (⊤ : PartENat).Dom
+
+-- variable {n : ℕ}
+-- #synth Decidable (n : PartENat).Dom
+-- end Test
+-- ```
+
+@[simp]
+theorem toWithTop_lt {x y : PartENat} [Decidable x.Dom] [Decidable y.Dom] :
+    toWithTop x < toWithTop y ↔ x < y :=
+  lt_iff_lt_of_le_iff_le toWithTop_le
+#align part_enat.to_with_top_lt PartENat.toWithTop_lt
+
+end WithTop
+
+-- Porting note : TODO should this be either backported or fix `withTopEquiv`
+-- in another way?
+
+-- Porting note : new, extracted from `withTopEquiv`.
+/-- Coersion from `ℕ∞` to `PartENat`. -/
+@[coe]
+def ofENat : ℕ∞ → PartENat :=
+  fun x => match x with
+  | Option.none => none
+  | Option.some n => some n
```

Comment:
I extracted this from `withTopEquiv` below because I had to deal with some def-eq problems. This should be looked at properly, I am not sure this is the best thing to do.

Would it need backport to mathlib3?